### PR TITLE
Add Repo toolchain support for the musl-repo target in the clang Driver.

### PR DESCRIPTION
**Pull request by MaggieYingYi**
After this change, the repo toolchain will be only used when targeting on the x86_64-pc-linux-musl-repo target.

### CODE REVIEW

**Comment by paulhuggett** [comment]
Could you elaborate on your description, please? I need to properly understand what your change does and, possibly more importantly, why and what the thinking behind it was.

**Comment by paulhuggett** [comment]
> - If the triple is x86_64-pc-linux-musl-repo [snip] musl libc based C/C++ libraries will be used.
> - If the triple is x86_64-pc-linux-gnu-repo, [snip] glibc is used by default.

All of our documentation currently uses x86_64-pc-linux-repo which is neither of the above. What happens in this case?

Your description implies that the _only_ difference between the behaviour of the two triples is the selection of the C library. Is that the case?

You list a lot of existing commits to support musl libc. Can we cherry pick some of those to support it rather than adding our own toolchain?

I assume that rld is used as the linker in all cases now but that doesn’t seem to be what the code is doing. (To use a traditional linker, you have to also run repo2obj, but there’s no mention of that). It also appears to be passing a lot of switches that rld doesn’t support (and that I would rather not support). What’s the story here?

You seem to have code for two different C++ standard libraries, are we planning to recompile both of them? (I think that one is probably sufficient.)

**Comment by MaggieYingYi** [comment]
>
> All of our documentation currently uses x86_64-pc-linux-repo which is neither of the above. What happens in this case?
>

It will keep the same behaviour as before, which will use the Linux toolchain. The Linux toolchain class is derived from Generic_ELF class, which is derived from Generic_GCC class. Generic_GCC is a tool chain using the `gcc` command to perform all sub-commands.

The RepoToolChain is only used when targeting on the musl environment and the repo ticket format (triple:
xxx-xxx-xxx-musl-repo). I have summarised the invoked toolchain for some triples, please see the table for the details.

| Triple | Invoked ToolChain before the commit | Invoked ToolChain after the commit | Behaviour Change? |
| -- | -- | -- | -- |
| x86_64-pc-linux-repo | Linux ToolChain | Linux ToolChain | NO |
| x86_64-pc-linux-musl-repo | Linux   ToolChain | RepoToolChain | YES |
| x86_64-pc-linux-unknow-repo | Linux ToolChain | Linux ToolChain | NO |
| x86_64-pc-unknow-repo | Generic_GCC | Generic_GCC | NO |
| x86_64-pc-unknow-musl-repo | Generic_GCC | RepoToolChain | YES |

> 
> Your description implies that the _only_ difference between the behaviour of the two triples is the selection of the C library. Is that the case?
>
Not really, the compiler, linker, included the default libraries (both C/C++ libraries) and the included header path might be different as well. The detailed difference could be found in Generic_GCC class and RepoToolChain class.

>
> You list a lot of existing commits to support musl libc. Can we cherry pick some of those to support it rather than adding our own toolchain?
>
All these commits are in the llvm-project-prepo. I have used the existing musl environment and add the repo toolchain to invoke musl libc based libraries.

> 
> I assume that rld is used as the linker in all cases now but that doesn’t seem to be what the code is doing. To use a traditional linker, you would have to also run repo2obj, but there’s no mention of that. It also appears to be passing a lot of switches that rld doesn’t support (and that I would rather not support). What’s the story here?
>
Yes, the rld is a default linker when targeting on musl-repo. As you mentioned, rld does support a lot of options and archive library yet, therefore, I only check that the clang driver includes the correct header path and default C/C++ libraries by passing `-###`.

> 
> You seem to have code for two different C++ standard libraries, are we planning to recompile both of them? (I think that one is probably sufficient.)
>
I agree with you – one is sufficient. This commit is the first step of adding repo toolchain to the clang driver. As mentioned before, it will not affect the existing repo tests which use the triple x86_64-pc-linux-repo (not use musl environment). Once rld supports the archive library and is ready to use, I will do a follow-up commit to make repo only use the musl libc based C/C++ libraries.

**Comment by paulhuggett** [comment]
> the compiler, linker, included the default libraries (both C/C++ libraries) and the included header path might be different as well

I understand a change to include paths and libraries, but why would changing the standard C library affect the compiler and/or linker being used? These are surely independent of one another.

> All these commits are in the llvm-project-prepo. I have used the existing musl environment and add the repo toolchain to invoke musl libc based libraries.

Do you mean that these commits predate our fork and are therefore already present in the llvm-project-prepo repository? Is there nothing in any of these commits that can be reused to reduce the need for more new code (e.g. setting include paths and default libraries)?

> rld does support a lot of options and archive library yet, therefore, I only check that the clang driver includes the correct header path and default C/C++ libraries by passing -###.

Repo.cpp/`constructRepoLinkArgs()` appears to add a lot of linker switches. Checking the compiler switches may not be sufficient. I do not plan to add rld support for most of the linker switches that this code appears to use.

> I agree with you – one is sufficient

Okay… if that’s the case, I don’t understand why is there new code which seems to supporting multiple C++ libraries (e.g. `RepoToolChain::GetCXXStdlibType(const ArgList &Args)`. Can you explain further?

**Comment by MaggieYingYi** [comment]
> 
> I understand a change to include paths and libraries, but why would changing the standard C library affect the compiler and/or linker being used? These are surely independent of one another.
> 
Sorry, I didn’t say clearly. The C library should not affect the compiler/linker being used. However, the clang driver might invoke the different linker which is decided by the default linker defined in the Toolchain. With this first commit, `rld` is used as the default linker when targeting on the musl-repo. Otherwise, `gnu ld` is used as the default linker in order to pass all existing repo tests. Once rld supports the archive library and is ready to use, we will use the rld as the default linker when targeting on Repo.

> 
> Do you mean that these commits predate our fork and are therefore already present in the llvm-project-prepo repository?
>
Yes, you are right. These commits exist in the llvm-project 11.0 release.

>
> Is there nothing in any of these commits that can be reused to reduce the need for more new code (e.g. setting include paths and default libraries)?
> 
I checked all the commits. I don’t think we could reuse any these commits. In the Toolchain file, the commit uses the musl dynamic library, “/lib/ld-musl-ArchName.so.1”. We are using the static musl libc library, therefore we can’t use the default musl library. For the hexgon target, the include paths and libraries are added in its own toolchain files ( lang/lib/Driver/ToolChains/Hexagon.cpp(.h). I also follows the similar way to add our repo toolchain. 

> 
> Repo.cpp/`constructRepoLinkArgs()` appears to add a lot of linker switches. Checking the compiler switches may not be sufficient. I do not plan to add rld support for most of the linker switches that this code appears to use.
> 
Thanks for letting me know that. I have removed unsupported rld switches (commit: 75c3bc3f). I only keep the linker option -static and -shared to switch between static link and dynamic link. Let me know if these two options are not needed as well. I have checked the invoked linker command from clang driver, the unsupported switches have been removed.

> 
> Okay… if that’s the case, I don’t understand why is there new code which seems to supporting multiple C++ libraries (e.g. `RepoToolChain::GetCXXStdlibType(const ArgList &Args)`. Can you explain further?
>
Good point. I have fixed in the commit 75c3bc3f.

**Comment by paulhuggett** [comment]
> The C library should not affect the compiler/linker being used. [snip] With this first commit, rld is used as the default linker when targeting on the musl-repo. Otherwise, gnu ld is used as the default linker in order to pass all existing repo tests

Here it seems that changing the C library _does_ affect the linker being used! I‘m not sure about this reasoning.

It seems strange to me that requesting ‘-repo’ format output can result in a linker being run that cannot process repo ticket files. If you want object files, use repo2obj to convert the tickets but then we’re no longer using repo.

IMO, existing tests that use rely on an object-file linker should not be using the -repo triple to link.

What happens if ld is not installed (e.g. inside our docker image)?

>  I only keep the linker option -static and -shared to switch between static link and dynamic link. Let me know if these two options are not needed as well. 

See [this file](https://github.com/SNSystems/llvm-project-prepo/blob/master/rld/tools/rld/rld.cpp). rld currently supports static linking only, so no -static or -shared.

**Comment by MaggieYingYi** [comment]
> 
> 
> Here it seems that changing the C library _does_ affect the linker being used! I‘m not sure about this reasoning.
> 
Changing the C library affect the invoked different ToolChain defined in the ` clang/lib/Driver/Driver.cpp` file. In that file, I modified the Driver to invoke the RepoToolChain when targeting on the musl environment and the repo ticket file format (for example, triple likes xxx-xxx-xxx-musl-repo). When targeting on the Linux and the repo ticket file but not on musl environment (for example, triple likes xxx-xxx-linux-repo), the Linux ToolChain will invoke. The function `const ToolChain &Driver::getToolChain` defined in the Driver.cpp file explained the reason.

>
> It seems strange to me that requesting ‘-repo’ format output can result in a linker being run that cannot process repo ticket files. If you want object files, use repo2obj to convert the tickets but then we’re no longer using repo. IMO, existing tests that rely on an object-file linker should not be using the -repo triple to link.
>
Sorry, I am not sure I understand your question. The new test only checks the compiler and linker commands by passing `-###`. There is no compilation and linking happen.

> 
> What happens if ld is not installed (e.g. inside our docker image)?
> 
Again, since the new test only shows the invoked commands by Driver and not really runs the linking command. Therefore, still pass in the Docker image.
> 
> See [this file](https://github.com/SNSystems/llvm-project-prepo/blob/master/rld/tools/rld/rld.cpp). rld currently supports static linking only, so no -static or -shared.
>
Removed in the commit https://github.com/SNSystems/llvm-project-prepo/pull/158/commits/d710f4a969acc1772056152c6749f76d6fffbe05.

**Comment by paulhuggett** [comment]
> Sorry, I am not sure I understand your question. The new test only checks the compiler and linker commands by passing `-###`.

It wasn’t a question! Let me try to put it another way. Consider:

1. The user installs glibc.
1. Compile: `clang -c -target x86_64-pc-linux-gnu-repo -o a.o c.o`.
1. Link: `clang -target x86_64-pc-linux-gnu-repo a.o`

   This will fail in the docker image because ld is not installed. 

1. The user installs ld.
1. 2nd attempt to link: `clang -target x86_64-pc-linux-gnu-repo a.o`

    This will fail because ld does not support ticket files. The _only_ way to make the link step work with ld is to convert the ticket file to ELF using repo2obj. 
  
    Once this has happened, we're no longer using repo output. The -repo triple should not then be used to link.

1. Convert to ELF: `repo2obj -o a.o.elf a.o`.
1. 3rd attempt to link: `clang a.o.elf`.

   Success. To link with repo2obj output, the ‘-repo’ triple should not be used.
   
To summarize: 
- Changing the C library shouldn’t change the linker being used. Using the repo format should always run rld.
- Specifying the -repo file format in the triple shouldn’t result in an ELF linker being run because this won’t be able to link the repo ticket files.

> the new test only shows the invoked commands by Driver and not really runs the linking command.

I’m not concerned about the new test.

**Comment by MaggieYingYi** [comment]
Thanks Paul for the detailed explanation.

I have updated the patch in the commit 75d4d88.

**Comment by MaggieYingYi** [comment]
Hi Paul,

I have updated the patch following your suggestions:

1) Renamed `RepoToolChain` to `RepoMuslToolChain`.
2) Reverted gnu toolchain (gnul.cpp) changes.
3) Change the PIC and PIE default value to false.
4) Updated the test of repo-toolchain-linux-musl.c to remove the -gnu-* tests.

Thanks,

File: clang/lib/Driver/ToolChains/Repo.cpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+  return;
+}
+
+void repo::Linker::ConstructJob(Compilation &C, const JobAction &JA,
+                                const InputInfo &Output,
+                                const InputInfoList &Inputs,
+                                const ArgList &Args,
+                                const char *LinkingOutput) const {
+  auto &RTC =
+      static_cast<const toolchains::RepoMuslToolChain &>(getToolChain());
```
Comment (paulhuggett) [inline clang/lib/Driver/ToolChains/Repo.cpp:94]: I'm confused by the role of the repo::Linker class. It seems to be used to describe rld which would mean that it is used by any triple that wants to link from the repo. However, here we seem to require that the triple is *-musl-repo.  Can you explain? (Perhaps add comments next to the class declaration to describe it.)

File: clang/lib/Driver/ToolChains/Repo.cpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+  //----------------------------------------------------------------------------
+  // Libraries
+  //----------------------------------------------------------------------------
+  if (!Args.hasArg(options::OPT_nostdlib, options::OPT_nodefaultlibs)) {
+    CmdArgs.push_back("-lclang_rt.builtins-x86_64");
+  }
+
+  if (D.CCCIsCXX()) {
+    if (RTC.ShouldLinkCXXStdlib(Args))
```
Comment (paulhuggett) [inline clang/lib/Driver/ToolChains/Repo.cpp:81]: Combine the two if statements:
```cxx
if (D.CCCIsCXX() && RTC.ShouldLinkCXXStdlib(Args)) {}
```

File: clang/lib/Driver/ToolChains/Repo.cpp [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+  //----------------------------------------------------------------------------
+  // Library Search Paths
+  //----------------------------------------------------------------------------
+  CmdArgs.push_back(Args.MakeArgString(StringRef("-L") + MuslRoot + "/lib"));
+
+  if (RTC.getVFS().exists(CRTPath))
+    CmdArgs.push_back(Args.MakeArgString("-L" + CRTPath));
+
+  const ToolChain::path_list &LibPaths = RTC.getFilePaths();
```
Comment (paulhuggett) [inline clang/lib/Driver/ToolChains/Repo.cpp:61]: Do we need this variable? It is only used once.

File: clang/lib/Driver/ToolChains/Repo.cpp [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+  const Driver &D = RTC.getDriver();
+
+  //----------------------------------------------------------------------------
+  //
+  //----------------------------------------------------------------------------
+  CmdArgs.push_back("-o");
+  CmdArgs.push_back(Output.getFilename());
+
+  std::string MuslRoot = D.SysRoot.empty() ? "/usr/local/musl" : D.SysRoot;
```
Comment (paulhuggett) [inline clang/lib/Driver/ToolChains/Repo.cpp:36]: Make this variable const?

File: clang/lib/Driver/ToolChains/Repo.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                    const InputInfo &Output, const InputInfoList &Inputs,
+                    const llvm::opt::ArgList &TCArgs,
+                    const char *LinkingOutput) const override;
+};
+} // namespace repo
+} // namespace tools
+
+namespace toolchains {
+
+class LLVM_LIBRARY_VISIBILITY RepoMuslToolChain : public ToolChain {
```
Comment (paulhuggett) [inline clang/lib/Driver/ToolChains/Repo.h:38]: I think that this class needs some explanation along the lines of what you told me during our phone call. Something to explain why "repo" and "musl" are bound together even though the only "repo" change is to use rld rather than an ELF linker.

File: clang/test/Driver/repo-toolchain-linux-musl.c [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+// CHECK004-NOT:   -lclang_rt.builtins-x86_64
+
+// -----------------------------------------------------------------------------
+// Checking the linked objects and libraries
+// Not Passing -fno-use-init-array when musl is selected
+// -----------------------------------------------------------------------------
+// RUN: %clang -### -target x86_64-pc-linux-musl-repo %s 2>&1 \
+// RUN:   | FileCheck -check-prefix=CHECK005 %s
+
+// RUN: %clang -### -target x86_64-pc-linux-gnu-repo %s 2>&1 \
```
Comment (paulhuggett) [inline clang/test/Driver/repo-toolchain-linux-musl.c:163]: Weren't you going to move the *-gnu-repo tests out of the *-musl.c source file? They don't seem to fit here.

Comment (MaggieYingYi) [reply clang/test/Driver/repo-toolchain-linux-musl.c]: Removed in the commit [0a629b99](https://github.com/SNSystems/llvm-project-prepo/pull/158/commits/0a629b99bb27274ce786c619db54e1a59d902947).

File: clang/lib/Driver/ToolChains/Repo.h (hunk 5 above)
Comment (MaggieYingYi) [reply clang/lib/Driver/ToolChains/Repo.h:47]: Added in the commit [0a629b99](https://github.com/SNSystems/llvm-project-prepo/pull/158/commits/0a629b99bb27274ce786c619db54e1a59d902947).

File: clang/lib/Driver/ToolChains/Repo.cpp (hunk 3 above)
Comment (MaggieYingYi) [reply clang/lib/Driver/ToolChains/Repo.cpp]: Fixed in the commit [0a629b99](https://github.com/SNSystems/llvm-project-prepo/pull/158/commits/0a629b99bb27274ce786c619db54e1a59d902947).

File: clang/lib/Driver/ToolChains/Repo.cpp (hunk 2 above)
Comment (MaggieYingYi) [reply clang/lib/Driver/ToolChains/Repo.cpp]: Fixed in the commit [0a629b99](https://github.com/SNSystems/llvm-project-prepo/pull/158/commits/0a629b99bb27274ce786c619db54e1a59d902947).

File: clang/lib/Driver/ToolChains/Repo.cpp (hunk 1 above)
Comment (MaggieYingYi) [reply clang/lib/Driver/ToolChains/Repo.cpp]: Good point. I have modified the code and added comments in the commit [0a629b99](https://github.com/SNSystems/llvm-project-prepo/pull/158/commits/0a629b99bb27274ce786c619db54e1a59d902947).

File: clang/lib/Driver/ToolChains/Repo.cpp [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+  const Driver &D = getDriver();
+
+  if (DriverArgs.hasArg(options::OPT_nostdlibinc))
+    return;
+
+  if (!DriverArgs.hasArg(options::OPT_nostdinc) &&
+      !DriverArgs.hasArg(options::OPT_nostdlibinc) &&
+      !DriverArgs.hasArg(options::OPT_nobuiltininc) &&
+      !DriverArgs.hasArg(options::OPT_nostdincxx)) {
+    // On musl-repo, libc++ is installed alongside the compiler in
```
[the remark this answers is not in the document]
Comment (MaggieYingYi) [reply clang/lib/Driver/ToolChains/Repo.cpp]: Yes, changed `musl-repo` to `repo` in the commit [0a629b99](https://github.com/SNSystems/llvm-project-prepo/pull/158/commits/0a629b99bb27274ce786c619db54e1a59d902947).

File: clang/lib/Driver/ToolChains/Repo.cpp [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+  if (!Args.hasArg(options::OPT_nostdlib, options::OPT_nodefaultlibs)) {
+    CmdArgs.push_back("-lclang_rt.builtins-x86_64");
+  }
+
+  if (D.CCCIsCXX()) {
+    if (RTC.ShouldLinkCXXStdlib(Args))
+      RTC.AddCXXStdlibLibArgs(Args, CmdArgs);
+  }
+
+  return;
```
[the remark this answers is not in the document]
Comment (MaggieYingYi) [reply clang/lib/Driver/ToolChains/Repo.cpp]: Removed in the commit [0a629b99](https://github.com/SNSystems/llvm-project-prepo/pull/158/commits/0a629b99bb27274ce786c619db54e1a59d902947).

File: clang/lib/Driver/ToolChains/Repo.cpp [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+  std::string MuslRoot = D.SysRoot.empty() ? "/usr/local/musl" : D.SysRoot;
+  if (!Args.hasArg(options::OPT_nostartfiles, options::OPT_nostdlib)) {
+    CmdArgs.push_back(Args.MakeArgString(MuslRoot + "/lib/crt1.t"));
+    CmdArgs.push_back(Args.MakeArgString(MuslRoot + "/lib/crt1_asm.t"));
+    CmdArgs.push_back(Args.MakeArgString(MuslRoot + "/lib/libc_repo.a"));
+  }
+
+  auto CRTPath = RTC.getCompilerRTPath();
+  if (!RTC.getVFS().exists(CRTPath)) // Build compiler-rt as stand-alone
+    CRTPath = "/usr/lib/linux";
```
[the remark this answers is not in the document]
Comment (MaggieYingYi) [reply clang/lib/Driver/ToolChains/Repo.cpp]: Fixed in the commit [0a629b99](https://github.com/SNSystems/llvm-project-prepo/pull/158/commits/0a629b99bb27274ce786c619db54e1a59d902947).

File: clang/lib/Driver/ToolChain.cpp [hunk 10]
```diff
@@ -555,6 +555,13 @@ std::string ToolChain::GetLinkerPath() const {
     // second-guess that.
     if (llvm::sys::fs::can_execute(UseLinker))
       return std::string(UseLinker);
+  } else if (Triple.isOSBinFormatRepo()) {
+    llvm::SmallString<8> LinkerName;
+    LinkerName.append("rld");
```
[the remark this answers is not in the document]
Comment (MaggieYingYi) [reply clang/lib/Driver/ToolChain.cpp]: Removed in the commit [0a629b99](https://github.com/SNSystems/llvm-project-prepo/pull/158/commits/0a629b99bb27274ce786c619db54e1a59d902947).

File: clang/lib/Driver/ToolChains/Repo.h [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+} // namespace tools
+
+namespace toolchains {
+
+/// RepoMuslToolChain - A tool chain using the repo tools and repo-based
+/// musl-libc and llvm libc++ libraries to perform compilation and linking
+/// commands. Currently does not support repo-based glibc library but
+/// compilation is fine.
+/// Since the *-musl-* toolchain doesn't support in the upstream LLVM, we need
+/// to add this class when targetting on repo and using musl libc linrary. We
```
Comment (paulhuggett) [inline clang/lib/Driver/ToolChains/Repo.h:44]: > Since the *-musl-* toolchain doesn't support in the upstream LLVM

Should be ”…toolchain isn’t supported by upstream…”

> when targetting on repo and using musl libc linrary

Should be ”when targeting the repo and using the musl libc library”.

File: clang/lib/Driver/ToolChains/Repo.h [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+/// could add the musl toolchain support (like gnu toolchain).
+/// TODO: support musl toolcahin in upstream LLVM.
```
Comment (paulhuggett) [inline clang/lib/Driver/ToolChains/Repo.h:46]: Sounds like you’re offering to do the work: I wouldn’t say that if I were you!

File: clang/lib/Driver/ToolChains/Repo.cpp [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+
+  // If compiler-rt is built inside of LLVM project, we could find its path
+  // using getCompilerRTPath function. If build compiler-rt as stand-alone
+  // project and give --sysroot=preference, use sysroot as the libary root.
+  auto CRTPath = RTC.getCompilerRTPath();
+  if (!RTC.getVFS().exists(CRTPath))
+    CRTPath = D.SysRoot.empty() ? "/usrlib/linux" : D.SysRoot + "/lib/linux/";
+  if (!Args.hasArg(options::OPT_nostartfiles, options::OPT_nostdlib)) {
+    CmdArgs.push_back(
+        Args.MakeArgString(CRTPath + "/clang_rt.crtbegin-x86_64.o"));
```
Comment (paulhuggett) [inline clang/lib/Driver/ToolChains/Repo.cpp:50]: Elsewhere, the code uses `llvm::sys::path::append()` to join path strings, but here (as well as in a few other places) we just concatenate the strings. This seems rather inconsistent.

File: clang/lib/Driver/ToolChains/Repo.cpp [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+  if (!D.SysRoot.empty()) {
+    SmallString<128> P(D.SysRoot);
+    llvm::sys::path::append(P, "lib");
+    getFilePaths().push_back(std::string(P.str()));
+  }
+
+  getFilePaths().push_back(D.Dir + "/../lib");
+}
+
+RepoMuslToolChain::~RepoMuslToolChain() {}
```
Comment (paulhuggett) [inline clang/lib/Driver/ToolChains/Repo.cpp:126]: Change `{}` to `= default;`.

File: clang/lib/Driver/ToolChains/Repo.cpp [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+ToolChain::CXXStdlibType
+RepoMuslToolChain::GetCXXStdlibType(const ArgList &Args) const {
+  Arg *A = Args.getLastArg(options::OPT_stdlib_EQ);
+  if (A) {
+    StringRef Value = A->getValue();
+    if (Value == "libstdc++")
+      return ToolChain::CST_Libstdcxx;
+    else if (Value == "libc++")
+      return ToolChain::CST_Libcxx;
+    else
```
[the remark this answers is not in the document]
Comment (MaggieYingYi) [reply clang/lib/Driver/ToolChains/Repo.cpp]: Fixed it  in the commit [7c20e553](https://github.com/SNSystems/llvm-project-prepo/pull/158/commits/7c20e55309a449948d30e3fbcc7ffe4ed3361198).

File: clang/lib/Driver/ToolChains/Repo.cpp [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+ToolChain::CXXStdlibType
+RepoMuslToolChain::GetCXXStdlibType(const ArgList &Args) const {
+  Arg *A = Args.getLastArg(options::OPT_stdlib_EQ);
+  if (A) {
+    StringRef Value = A->getValue();
+    if (Value == "libstdc++")
+      return ToolChain::CST_Libstdcxx;
+    else if (Value == "libc++")
```
[the remark this answers is not in the document]
Comment (MaggieYingYi) [reply clang/lib/Driver/ToolChains/Repo.cpp]: Fixed it  in the commit [7c20e553](https://github.com/SNSystems/llvm-project-prepo/pull/158/commits/7c20e55309a449948d30e3fbcc7ffe4ed3361198).

File: clang/lib/Driver/ToolChains/Repo.h (hunk 11 above)
Comment (MaggieYingYi) [reply clang/lib/Driver/ToolChains/Repo.h]: Fixed it in the commit [7c20e553](https://github.com/SNSystems/llvm-project-prepo/pull/158/commits/7c20e55309a449948d30e3fbcc7ffe4ed3361198).

File: clang/lib/Driver/ToolChains/Repo.h (hunk 12 above)
Comment (MaggieYingYi) [reply clang/lib/Driver/ToolChains/Repo.h]: Removed `TODO` in the commit [7c20e553](https://github.com/SNSystems/llvm-project-prepo/pull/158/commits/7c20e55309a449948d30e3fbcc7ffe4ed3361198).

File: clang/lib/Driver/ToolChains/Repo.cpp (hunk 13 above)
Comment (MaggieYingYi) [reply clang/lib/Driver/ToolChains/Repo.cpp]: I have changed to use the `llvm::sys::path::append()`. However, for the string passing to `Args.MakeArgString` function, I have to add two variables `CRTBegin` and `CRTEnd`. These two variables are only used one time. For the arguments of `Args.MakeArgString`, should we still used concatenate the strings (like other LLVM files/codes)? 

>
> as well as in a few other places.
>
Other places are mainly for the arguments of `Args.MakeArgString`, I still used concatenate the strings (like other LLVM files/codes).

File: clang/lib/Driver/ToolChains/Repo.cpp (hunk 14 above)
Comment (MaggieYingYi) [reply clang/lib/Driver/ToolChains/Repo.cpp]: Fixed it in the commit [7c20e553](https://github.com/SNSystems/llvm-project-prepo/pull/158/commits/7c20e55309a449948d30e3fbcc7ffe4ed3361198).

File: clang/test/Driver/repo-toolchain-linux-musl.c [hunk 17]
```diff
@@ -170,40 +188,40 @@
...
 // RUN: %clang -### -target x86_64-pc-linux-gnu %s 2>&1 \
-// RUN:   | FileCheck -check-prefix=CHECK007 %s
+// RUN:   | FileCheck -check-prefix=CHECK008 %s
 
 // RUN: %clang -### -target x86_64-pc-linux %s 2>&1 \
-// RUN:   | FileCheck -check-prefix=CHECK007 %s
+// RUN:   | FileCheck -check-prefix=CHECK008 %s
 
-// CHECK007:      "{{.*}}ld"
+// CHECK008:      "{{.*}}ld"
```
Comment (paulhuggett) [inline clang/test/Driver/repo-toolchain-linux-musl.c:212]: Won't this test pass for any string ending in “ld” (which means most ELF linkers)‽ Should it end with ”/ld”?

**Review by paulhuggett — APPROVED** [review]

File: clang/test/Driver/repo-toolchain-linux-musl.c (hunk 17 above)
Comment (MaggieYingYi) [reply clang/test/Driver/repo-toolchain-linux-musl.c:212]: On windows, the Driver can’t find the `ld` path, it will use `ld`. For example,
```
"ld" "-pie" "--eh-frame-hdr" "-m" "elf_x86_64" …
```
Therefore, we can’t use `/ld` on Windows.
